### PR TITLE
feat(eslint-plugin): add extension rule `space-before-blocks` (#1606)

### LINKED ISSUE #1606: Missing spaces around name interface and type
[Repository with example.](https://gitlab.com/valentineus/typescript-eslint-example)

### Instruction

1. Clone a repository:

```bash
git clone https://gitlab.com/valentineus/typescript-eslint-example.git
```

2. Install dependencies:

```bash
npm install
```

3. Run a test:

```bash
npm run lint
```

Code to check:

```typescript
interface ITesstInterfase{
    key: string;
}

type TestType=() => void;

const variable1: TestType = () => {

    console.log("Hello, World!");

};

variable1();
```

Enabled all rules `eslint` and `@typescript-eslint/eslint-plugin`.

On line 1, missing spaces around interface `ITesstInterfase`.
On line 5, missing spaces around sign `=`.

Are there any rules governing such cases?

|            **package**             | **version** |
| ---------------------------------- | ----------- |
| `@typescript-eslint/eslint-plugin` | `2.19.2`    |
| `@typescript-eslint/parser`        | `2.19.2`    |
| `eslint`                           | `6.8.0`     |

### PR DESCRIPTION
**Pull request by FDIM**
## PR Checklist

-   [x] Addresses an existing issue: fixes #1606 
-   [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
-   [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/master/CONTRIBUTING.md) were taken

## Overview

I've added an extension rule for `space-before-blocks` that ensures that space before curly brace is enforced for enums and interfaces. 

~~I'll add rule documentation file a bit later and in the meantime, feedback on PR is welcome as I'm not too familiar with this code base. Most specifically if there is any better way of handling schema without copying & extending.~~ (PR is ready)

Impressive local setup, even debugging via vscode just worked! 

EDIT: and even more impressive setup in CI, unit test for missing doc file was a nice surprise.

### CODE REVIEW

File: packages/eslint-plugin/typings/eslint-rules.d.ts
Comment (FDIM) [inline packages/eslint-plugin/typings/eslint-rules.d.ts:839]: Will update type to match base rule, this one does not match 1-1)

File: packages/eslint-plugin/src/rules/space-before-blocks.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/space-before-blocks.ts:51]:
```suggestion
      node: TSESTree.Token | TSESTree.TSInterfaceBody,
```

`TSESTree.PunctuatorToken` is a part of `TSESTree.Token`. https://github.com/typescript-eslint/typescript-eslint/issues/2738

File: packages/eslint-plugin/src/rules/space-before-blocks.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/space-before-blocks.ts:56]: Casting a non-token node `as TSESTree.Token` is unsafe. Since it looks like `util.isTokenOnSameLine` works if the second argument is a regular node the right approach would be to make it `right: TSESTree.Node | TSESTree.Token`.

File: packages/eslint-plugin/src/rules/space-before-blocks.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/space-before-blocks.ts:58]: Err, `isSpaceBetweenTokens` is marked as `@deprecated in favor of isSpaceBetween`. Have you tried using that instead?

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
LGTM in general -- thanks for adding this in @FDIM! ❤️‍🔥 

I'm not particularly familiar with extension rules yet so I'll defer to @bradzacher or another maintainer.

Just requesting changes on a few type system and API usage fixups.

File: packages/eslint-plugin/src/rules/space-before-blocks.ts
Comment (bradzacher) [reply packages/eslint-plugin/src/rules/space-before-blocks.ts:25]: This would have been a side-effect of #3844.
We could definitely improve `getESLintCodeRule` to make these mutable!

---

This is an unfortunate problem that arrises because we support ESLint v6. Many rules didn't have `messages` defined until ESLint v7 (when myself and another contributor added them to all rules).

**Review by bradzacher — COMMENTED** [review]
all josh's comments are on point!
otherwise looks to be extended appropriately

File: packages/eslint-plugin/src/rules/space-before-blocks.ts
Comment (FDIM) [reply packages/eslint-plugin/src/rules/space-before-blocks.ts:37]: I found this and the always/never pattern a bit off but left it as because that's how it is in the original rule. Thought it would have been simpler to see differences. I'll update :)

File: packages/eslint-plugin/src/rules/space-before-blocks.ts
Comment (FDIM) [reply packages/eslint-plugin/src/rules/space-before-blocks.ts:58]: Replaced with `isSpaceBetween`, it is marked as optional though and available since 6.7. Should I use both then?
New method also doesn't require explicit cast, so that is nice.

Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/src/rules/space-before-blocks.ts]: Ah sorry for the delay @FDIM, missed your reply till now! Looking a bit closer, yes, `eslint-plugin` supports `"eslint": "^5.0.0 || ^6.0.0 || ^7.0.0"`. So we'll have to support both I think.

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Just the one thing on supporting older ESLint versions, otherwise 💯 !

Edit: and merging in from the `main` branch, if you don't mind 😄

**Comment by JoshuaKGoldberg** [comment]
Btw there's no need to force push. We squash PR commits. Force pushing makes it a little harder to review in some cases not hit by this PR for us.

**Comment by FDIM** [comment]
@JoshuaKGoldberg It should be ready again :D And sorry taking this long and for force pushing, got into this habit when working a small PRs for angular.

File: packages/eslint-plugin/src/rules/space-before-blocks.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/space-before-blocks.ts:1]: We renamed the package; `*/exprimental-utils` is now a thin wrapper.

```suggestion
import { TSESTree } from '@typescript-eslint/utils';
```

**Comment by JoshuaKGoldberg** [comment]
I've always been more of a React dev myself... 😜

File: packages/eslint-plugin/docs/rules/space-before-blocks.md
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/docs/rules/space-before-blocks.md:1]:
```suggestion
# space-before-blocks

Enforces consistent spacing before blocks.
```

https://github.com/typescript-eslint/typescript-eslint/runs/5175689231?check_suite_focus=true

We updated this format recently in #4367.

**Comment by FDIM** [comment]
@JoshuaKGoldberg few iterations later, everything should finally be good 😄

**Review by JoshuaKGoldberg — APPROVED** [review]
Super, thanks @FDIM! ✨

**Pull request merged**
